py/modmicropython: Expose repl_autocomplete as python function.

andrewleech opened the pull request:
### Summary

Exposes `mp_repl_autocomplete()` and `mp_hal_stdio_mode_raw()`/`mp_hal_stdio_mode_orig()` to Python as `micropython.repl_autocomplete()` and `micropython.stdio_mode_raw()` respectively.

`repl_autocomplete(line)` returns the completion suffix string, empty string for no match, or `None` when multiple candidates are printed. Gated on `MICROPY_HELPER_REPL` which is already enabled on most ports.

`stdio_mode_raw(enabled)` switches the terminal between raw and original mode. Gated behind a new `MICROPY_PY_MICROPYTHON_STDIO_RAW` config option, defaulting to off, enabled on unix.

Both are used by micropython/micropython-lib#1081 to bring aiorepl closer to feature parity with the native REPL — tab completion via `repl_autocomplete`, and proper terminal mode management via `stdio_mode_raw`.

### Testing

Unit tests added for both functions under `tests/micropython/`. A cmdline REPL test for `stdio_mode_raw` verifies actual terminal attribute changes via `termios`.

Tested on unix port.

### Generative AI

I used generative AI tools when creating this PR, but a human has checked the code and is responsible for the description above.

mattytrentini commented:
Adding autocomplete for aiorepl is _very_ desirable, and this is looking promising! Here are some of the tests I ran:

```python
>>>import micropython
>>>micropython.repl_autocomplete("impo")  # Should complete 'import'
'rt '
>>> class Foo:
...     def _bar():
...         pass
...     def alpha():
...         pass
>>> micropython.repl_autocomplete("f = Fo") # Should complete Foo
'o'
>>>f = Foo()
>>> micropython.repl_autocomplete("f.") # Should complete alpha (ignoring _bar)
'alpha'
```
Note that this is correct and consistent with MicroPython's built-in tab completion - but _differs_ in some cases to CPython, at least at v3.12. CPython will also supply parentheses, ie it would return `'o()'` to complete `f = Fo` and `'alpha('` for `f.`. That's not a problem with this PR, but should be resolved in `mp_repl_autocomplete`.

I also tested a module with private members and they were also correctly filtered out.

Looks good!

Josverl commented:
> but differs in some cases to CPython, at least at v3.12.

Perhaps that should _just_ be documented as such. I'd be fine with consistent behavior with the MicroPython repl.

dpgeorge commented on py/repl.c:
```diff
@@ -218,7 +218,7 @@ static void print_completions(const mp_print_t *print,
         size_t d_len;
         const char *d_str = (const char *)qstr_data(q, &d_len);
-        if (s_len <= d_len && strncmp(s_start, d_str, s_len) == 0) {
+        if (s_len <= d_len && strncmp(s_start, d_str, s_len) == 0 && d_str[0] != '_') {
```
Please break this out into a separate PR.  It'll need changes to the tests as well.

andrewleech replied:
Split out to https://github.com/micropython/micropython/pull/17108 with a unit test (which might need some more updating to ensure the test is consistent / works across ports etc?)

dpgeorge commented on py/modmicropython.c:
```diff
@@ -166,6 +167,18 @@ static mp_obj_t mp_micropython_schedule(mp_obj_t function, mp_obj_t arg) {
+    const char *str = mp_obj_str_get_data(cur_line, &str_len);
+
+    ssize_t compl_len = mp_repl_autocomplete(str, str_len, &mp_plat_print, &compl_str);
+    return (compl_len <= 0) ? mp_const_none : mp_obj_new_str_via_qstr(compl_str, compl_len);
```
If you make this `(compl_len < 0)` then the caller should be able to distinguish between the 3 distinct cases (no match, one match, many matches).

pi-anl committed:
py/modmicropython: Expose repl_autocomplete as python function.

Allows adding tab completion to custom REPLs such as aiorepl.

Returns a non-empty string (the completion suffix) on a unique match or
common prefix, an empty string when multiple candidates are printed to
stdout, or None when there is no match.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>
